File: 数据结构/ST表.py
class ST:
    def __init__(self, a, n):
        """
        a: 1-indexed 的数组
        """
        self.n = n
        self.m = n.bit_length()
        self.st = [[0] * self.m for _ in range(n + 1)]
        
        # 初始化长度为 1 的区间
        for i in range(1, n + 1):
            self.st[i][0] = a[i]
        
        # 构建 ST 表
        for j in range(1, self.m):
            for i in range(1, n - (1 << j) + 2):
                self.st[i][j] = max(self.st[i][j - 1], self.st[i + (1 << (j - 1))][j - 1])
    
    def query(self, l, r):
        """查询区间 [l, r] 的最大值"""
        k = (r - l + 1).bit_length() - 1
        return max(self.st[l][k], self.st[r - (1 << k) + 1][k])

File: 数据结构/ST表.cpp
struct ST {
    vector<vector<int>> st; // 从 i 开始 长度为 2^j 的区间的（max， min， gcd， &， |）
    ST(const vector<int>& a){
        int n = a.size() - 1; // 1-based
        st.assign(n + 1, vector<int>(log2(n) + 1));
        for (int i = 1; i <= n; ++i) st[i][0] = a[i];

        for (int j = 1; (1ll << j) <= n; ++j){
            for (int i = 1; i + (1ll << j) - 1 <= n; ++i){
                st[i][j] = max(st[i][j - 1], st[i + (1ll << (j - 1))][j - 1]);
            }
        }
    }

    int query(int l, int r){
        int k = log2(r - l + 1);
        return max(st[l][k], st[r - (1ll << k) + 1][k]);
    }
};


vector<vector<int>> st(n + 1, vector<int>(21)); // u点向上跳1<<i层的祖先节点
vector<int> dep(n + 1);
auto dfs = [&](auto && self, int u, int fa) -> void {
	dep[u] = dep[fa] + 1;
	st[u][0] = fa;
	for (int i = 1; i <= 20; ++i){
		st[u][i] = st[st[u][i - 1]][i - 1];
	}

	for (int v : g[u]){
		if (v == fa) continue;
		self(self, v, u);
	}
};

auto LCA = [&](int u, int v) -> int{
	if (dep[u] < dep[v]) swap(u, v);
	// u跳至与v同一层
	for (int i = 20; i >= 0; --i){
		if (dep[st[u][i]] >= dep[v]) u = st[u][i];
	}
	if (u == v) return v;
	// 一起跳至LCA的下一层
	for (int i = 20; i >= 0; --i){
		if (st[u][i] != st[v][i]) u = st[u][i], v = st[v][i];
	}
	return st[v][0];
};

dfs(dfs, rt, 0);
